plugin-react: maintainance

pengx17 opened the pull request:
What I did to continue the work on https://github.com/vitejs/vite/pull/3588
- update React related examples in the playground
- fix some build/plugin issues and make sure they can work in the playground
- add new playground/react-jsx-runtime and tests

pengx17 committed:
chore: update vite-dts

pengx17 committed:
build: use tsc instead of vite to build the plugin

pengx17 committed:
build: fix demo

pengx17 committed:
fix: automatic jsx runtime

pengx17 committed:
fix: esm/cjs build compatibility

aleclarson commented on packages/plugin-react/src/index.ts:
```diff
@@ -110,7 +118,10 @@ export default function viteReact(opts: Options = {}): PluginOption[] {
-              { runtime: 'automatic' }
+              {
+                runtime: 'automatic',
+                importSource: opts.jsxImportSource ?? 'react'
```
The `?? 'react'` part is unnecessary. Just for clarity?

pengx17 replied:
Yeah, it should be unnecessary. I will remove it.

aleclarson commented on packages/plugin-react/src/index.ts:
```diff
@@ -23,6 +23,14 @@ export interface Options {
+   * See https://babeljs.io/docs/en/babel-plugin-transform-react-jsx#importsource
+   * @default "react"
+   */
+  jsxImportSource?: string
```
Is there actually a use case for exposing this option?

pengx17 replied:
Yes. e.g., if I want to use `jsx` from `'@emotion/react'` instead from `'react'`, I will use
```js
    reactPlugin({
      jsxRuntime: 'automatic',
      jsxImportSource: '@emotion/react'
    })
```
see https://emotion.sh/docs/css-prop#:~:text=configuration-,babelrc,-presets

aleclarson commented on packages/plugin-react/src/index.ts:
```diff
@@ -184,10 +200,9 @@ export default function viteReact(opts: Options = {}): PluginOption[] {
       config.plugins.forEach(
         (plugin) =>
-          (plugin.name === 'react-refresh' ||
-            plugin.name === 'vite:react-jsx') &&
```
Should keep this check. It provides a smoother upgrade path for users of [vite-react-jsx](https://www.npmjs.com/package/vite-react-jsx).

pengx17 replied:
I just realize 'vite:react-jsx' here does not refer to the plugin in the current plugin, but refers to the one in your `vite-react-jsx` plugin.
I have changed 'vite:react-jsx' to 'vite:react-jsx-runtime-automic' to make it more clearer

aleclarson commented on packages/plugin-react/src/index.ts:
```diff
@@ -184,10 +200,9 @@ export default function viteReact(opts: Options = {}): PluginOption[] {
+          plugin.name === 'react-refresh' &&
           config.logger.warn(
-            `[@vitejs/plugin-react] This plugin conflicts with "${plugin.name}". Please remove it.`
+            `[@vitejs/react-refresh] This plugin conflicts with "${plugin.name}". Please remove it.`
```
The part between `[` and `]` is meant to be the plugin where the warning originated from. Maybe I should just rewrite the message to be clearer about what the user should do.

aleclarson commented on packages/plugin-react/src/index.ts:
```diff
@@ -174,6 +185,11 @@ export default function viteReact(opts: Options = {}): PluginOption[] {
+      // react/jsx-runtime is in CJS format
+      // we want to explicitly cast it to ESM
+      optimizeDeps: {
+        include: ['react/jsx-runtime']
```
You sure this is necessary? My project was working fine without it. 🤔

pengx17 replied:
I have tested with `packages/playground/react` which now enables `automic` runtime. Without explicitly optimize `react/jsx-runtime`, during runtime the browser tries to load `react/jsx-runtime` which is plain CJS with the following content:
```js
'use strict';

if (process.env.NODE_ENV === 'production') {
  module.exports = require('./cjs/react-jsx-runtime.production.min.js');
} else {
  module.exports = require('./cjs/react-jsx-runtime.development.js');
}
```

I think the issue is that during vite's import scanning phase this is not being seen. I am not sure if there is another way to include it into optimization instead of put it here?

aleclarson replied:
Oh it happens in development. That makes sense, as `vite-react-jsx` doesn't apply the transform in development, but this plugin does.

aleclarson commented:
Thanks for this. I'll rebase my PR to get rid of the commit noise.

Why was it necessary to stop using Vite to build the plugin?

pengx17 commented:
> Why was it necessary to stop using Vite to build the plugin?

When this plugin is built with vite, there is an issue that it will generate the following line in the output.

```js
var preloadHelper = require('vite/preload-helper');
// or
import { __vitePreload } from 'vite/preload-helper';
```

This `preloadHelper` is added by Vite's `buildImportAnalysisPlugin` when the source code has dynamic import, which are `import('@babel/plugin-transform-react-jsx-self')`, `import('@babel/plugin-transform-react-jsx-source')` etc.

However, this `vite/preload-helper` is not an actual module but something that is handled by `buildImportAnalysisPlugin.load`; also, it's implementation depends on DOM API.

Based on the above, I feel like maybe Vite is not a good build tool for this plugin yet, so I switched back to plain `tsc` instead.

pengx17 committed:
fix: fix review comments

aleclarson commented on packages/plugin-react/src/index.ts:
```diff
@@ -231,7 +232,8 @@ export default function viteReact(opts: Options = {}): PluginOption[] {
   const viteReactJsx: Plugin = {
-    name: 'vite:react-jsx',
+    // `alloc/vite-react-jsx` already takes `vite:react-jsx` name.
+    name: 'vite:react-jsx-runtime-automic',
```
Another option is to keep this `vite:react-jsx` and do the following check:
```js
(plugin !== viteReactJsx && plugin.name === 'vite:react-jsx')
```

pengx17 committed:
fix: a new way to check plugin conflicts

Commit:
chore: update lockfile

aleclarson commented on packages/playground/react/vite.config.ts:
```diff
@@ -0,0 +1,16 @@
+const config: import('vite').UserConfig = {
+  plugins: [
+    reactPlugin({
+      jsxRuntime: 'automatic'
```
Please duplicate `playground/react` to `playground/react-jsx-runtime` and use `jsxRuntime: "automatic"` in that playground only.

aleclarson replied:
We should probably test out the `babel` option in this playground.

```js
babel: {
  parserOpts: {
    plugins: ['decorators-legacy']
  }
}
```

Pick any parser plugin you like from this list:
https://babeljs.io/docs/en/babel-parser#ecmascript-proposalshttpsgithubcombabelproposals

pengx17 replied:
Hi Alec, I tried `pipelineOperator` parser, but it seems we cannot use the parser alone. This option only allow babel to be able to parse the file, but the output will stay unchanged, but later esbuild will then transpile the file and then esbuild will complain about pipeline syntax which is not supported by esbuild yet.
In the end, I used the plugin instead. I hope this is ok?

aleclarson commented:
Btw, you'll need to do `git reset --hard` since I rebased your PR.

pengx17 committed:
feat(plugin-react): add a new playground to test jsx-runtime automic

pengx17 commented:
@aleclarson hi, if you believe this is ready, you can merge it and I will continue work on the remaining task.

aleclarson commented:
@pengx17 Tests are not passing. Can you check it out?

pengx17 committed:
feat: import react also checks jsxInject

pengx17 committed:
build(plugin-react): add build-plugin-react to GitHub Actions

pengx17 committed:
chore: remove unused code

pengx17 committed:
fix: lint issues

pengx17 commented:
@aleclarson I checked the workflow and fixed an issue in https://github.com/aleclarson/vite/commit/0a7ad45ef39ae2f7c91142e257ce691c1c325334.
Also started gh actions inside of my fork here: https://github.com/pengx17/vite/pull/1, which looks like the tests run successfully now :)

aleclarson commented on packages/plugin-react/src/index.ts:
```diff
@@ -135,7 +141,13 @@ export default function viteReact(opts: Options = {}): PluginOption[] {
+            !isNodeModules &&
+            !(
+              importReactRegex.test(code) ||
+              importReactRegex.test(jsxInject ?? '')
```
We should also warn the user that `jsxInject` is unnecessary when using this plugin.

Something like this:
```ts
// in `configResolved` hook
logger = config.logger

// in `transform` hook
logger.warnOnce(
  '[vite-plugin-react] This plugin imports React for you automatically, ' +
    'so you can stop using `esbuild.jsxInject` for that purpose.'
)
```

aleclarson commented on packages/plugin-react/src/index.ts:
```diff
@@ -50,14 +50,20 @@ export default function viteReact(opts: Options = {}): PluginOption[] {
     name: 'vite:react-babel',
     enforce: 'pre',
+    configResolved(config) {
+      jsxInject = config.esbuild ? config.esbuild.jsxInject : undefined
```
Better use `config.esbuild?.jsxInject` here

pengx17 replied:
using `config.esbuild?.jsxInject` is a type error because `config.esbuild` can also be type of `false`.

aleclarson commented on packages/plugin-react/src/index.ts:
```diff
@@ -50,14 +50,20 @@ export default function viteReact(opts: Options = {}): PluginOption[] {
   const userParserPlugins =
     opts.parserPlugins || opts.babel?.parserOpts?.plugins || []
 
+  const importReactRegex = /(^|\n)import\s+(\*\s+as\s+)?React\s+/
```
Nice catch!

pengx17 committed:
fix: add a warning when using jsxInject to import React

aleclarson committed:
chore: fix `dev` script

aleclarson committed:
chore: various tweaks

aleclarson committed:
chore: remove src/index.d.ts

aleclarson commented on packages/plugin-react/src/index.ts:
```diff
@@ -256,3 +288,7 @@ viteReact.preambleCode = preambleCode
 }
+
+// overwrite for cjs require('...')() usage
+module.exports = viteReact
```
Are you sure this is necessary? What happens without it?

pengx17 replied:
The fix is brought from https://github.com/vitejs/vite/blob/main/packages/plugin-vue/src/index.ts#L190-L192

If this line is removed, when `plugin-react` is transpiled into commonjs module, the export is:

```js
exports.default = viteReact;
```

However, in Node if we do `require('@vitejs/plugin-react')`, the return value is `{default: viteReact}` instead of `viteReact`. Node does not convert `default` import into a namespace. 

Without this hack, when user `require` the plugin instead of using `import`, he will see:

```
error when starting dev server:
TypeError: reactPlugin is not a function
    at Object.<anonymous> (/Users/pengxiao/src/vite/packages/playground/react/vite.config.ts:15:13)
    at Module._compile (node:internal/modules/cjs/loader:1108:14)
```

aleclarson replied:
Have you done any testing in ESM environment?

Might need to wrap it in `if (typeof module !== 'undefined')`

pengx17 replied:
I just tested by turning on `"type": "module"` for playground/react, it looks like transpiling with TSC is not enough to let it be used in ESM environment. The error:
```
failed to load config from /Users/stb/Documents/GitHub/vite/packages/playground/react/vite.config.ts
error when starting dev server:
/Users/stb/Documents/GitHub/vite/packages/plugin-react/dist/esm/index.js:1
import * as babel from '@babel/core';
^^^^^^

SyntaxError: Cannot use import statement outside a module
```
It seems it is not trivial to fix it without a bundler. A related ref https://github.com/microsoft/TypeScript/issues/18442

I will try a similar approach in `plugin-vue` instead with esbuild.

aleclarson replied:
Okay, I guess it's fine we don't support ESM, since `plugin-vue` doesn't either.

pengx17 replied:
Yeah, I guess we can add esm built later if there is some real use cases.

aleclarson commented on packages/plugin-react/src/index.ts:
```diff
@@ -93,7 +108,7 @@ export default function viteReact(opts: Options = {}): PluginOption[] {
             useFastRefresh = true
             plugins.push([
-              await import('react-refresh/babel'),
+              await interopDefault(import('react-refresh/babel')),
```
I don't think this call is necessary. See [here](https://unpkg.com/browse/react-refresh@0.10.0/babel.js). This module uses CommonJS syntax.

aleclarson replied:
Hmm okay. Guess I'm wrong. Seems that the CommonJS module is coerced into an ES `Module` instance by dynamic import.

aleclarson replied:
Interestingly, I get this error when doing `import('react-refresh/babel')` in Node v14 REPL
```
Error [ERR_MODULE_NOT_FOUND]: Cannot find module '~/vite/node_modules/react-refresh/babel' imported from ~/vite/packages/plugin-react/repl
Did you mean to import react-refresh/babel.js?
```
..while `import('react-refresh/babel.js')` works fine.

aleclarson committed:
fix: use explicit extension with dynamic import

aleclarson commented on packages/playground/react-jsx-runtime/App.jsx:
```diff
@@ -0,0 +1,54 @@
+}
+
+// @ts-expect-error ts does not support pipeline operators yet
+const Counter = StatelessCounter |> withState |> ((_) => withWrapper(_, 'p'))
```
TypeScript is reporting `Expression expected. ts(1109)` on this line. Also, the optimizer is choking on this syntax when scanning for imports. We should use another plugin for now.

pengx17 replied:
Ops. I think this worked because of cache thus import analysis was skipped without choking. I will find one that is esbuild compatibility

pengx17 replied:
After some investigations, I think we cannot add parser/plugin for stage < 3 proposals in a vote playground, since stage < 3 proposal is not supported by both TS & esbuild, see a related issue https://github.com/vitejs/vite/issues/2349

pengx17 replied:
I updated the playground to demo how to enable `@emotion/react` with `@vite/plugin-react` to fix a wanted issue: https://github.com/vitejs/vite/issues/3276
Also added a test for it, which is similar to the original `playground/react`, without `file location metadata` check.
However, I am not sure what happened for this test, but the tests are now failing on windows.
https://github.com/pengx17/vite/runs/3150400557?check_suite_focus=true

It is reporting the following error when running on Windows/Node 12 for build test
```
    [vite]: Rollup failed to resolve import "D:aiteite
    ode_modules
    eactjsx-runtime.js" from "react/jsx-runtime".
    This is most likely unintended because it can break your application at runtime.
    If you do want to externalize this module explicitly add it to
    `build.rollupOptions.external`
```

Weird that this issue does not happen for other scenarios.

pengx17 replied:
The above error seems to be related to the code added in viteReactJsx plugin that causes issues on Windows. After a second look, I think viteReactJsx is not actually necessary and vite can successfully transform `react/jsx-runtime` into esm on the fly.

aleclarson replied:
Fixed with https://github.com/aleclarson/vite/pull/3/commits/e6bcfae5ed79b1ff74a9885230ab9df9c00dc263, I think

pengx17 replied:
Confirmed and the tests have passed

aleclarson reviewed (changes requested):
Also, it looks like `playground/react-jsx-runtime` has no `__tests__` folder. We should add some automated tests so the CI can catch any issues.

pengx17 committed:
build: also build-plugin-react at root's build

pengx17 committed:
build: use esbuild to bundle plugin-react

pengx17 committed:
fix: should not add legacy plugins to automatic runtime

pengx17 committed:
test: use emotion to demo react-jsx-runtime tests

pengx17 committed:
fix: move react/jsx-runtime optimization to vite:react-jsx

pengx17 commented on packages/plugin-react/src/index.ts:
```diff
@@ -68,23 +83,10 @@ export default function viteReact(opts: Options = {}): PluginOption[] {
-          // TODO: maybe we need to read tsconfig to determine parser plugins to
-          // enable here, but allowing decorators by default since it's very
-          // commonly used with TS.
-          parserPlugins.push('typescript', 'decorators-legacy')
```
`decorators-legacy` won't work for now anyway, so it is removed.

pengx17 committed:
fix: remove unnecessary react/jsx-runtime loader

aleclarson committed:
Revert "fix: remove unnecessary react/jsx-runtime loader"

This reverts commit f9006ba470ee20bd07cff76c5fdb01e56d3bccee.

aleclarson committed:
test: add react-switch to playground/react-jsx-runtime

This dependency is minified and commonjs, which will test the capabilities
of the `restoreJSX` function.

aleclarson committed:
fix: ensure node_modules is converted to automatic jsx-runtime

aleclarson committed:
fix: make vite:react-jsx work in Windows env

aleclarson committed:
test: fragment syntax

aleclarson commented:
Here's why the `vite:react-jsx` plugin is needed. 👇 

The `dist/assets/index.[hash].js` file **without** it: (notice the `jsxRuntime.exports.` spam)
```js
import { _ as _Switch, j as jsxRuntime, r as react, c as css, R as ReactDOM, a as React } from "./vendor.9e0ab00b.js";
const Switch = _Switch.default;
function Counter() {...}
function FragmentTest() {...}
function App() {
  return /* @__PURE__ */ jsxRuntime.exports.jsx("div", {
    className: "App",
    children: /* @__PURE__ */ jsxRuntime.exports.jsxs("header", {
      className: "App-header",
      children: [/* @__PURE__ */ jsxRuntime.exports.jsx("h1", {
        children: "Hello Vite + React + @emotion/react"
      }), /* @__PURE__ */ jsxRuntime.exports.jsx(FragmentTest, {}), /* @__PURE__ */ jsxRuntime.exports.jsxs("p", {
        children: ["Edit ", /* @__PURE__ */ jsxRuntime.exports.jsx("code", {
          children: "App.jsx"
        }), " and save to test HMR updates."]
      }), /* @__PURE__ */ jsxRuntime.exports.jsx("a", {
        className: "App-link",
        href: "https://reactjs.org",
        target: "_blank",
        rel: "noopener noreferrer",
        children: "Learn React"
      })]
    })
  });
}
ReactDOM.render(React.createElement(App), document.getElementById("app"));
```

Then **with** the plugin you thought was unnecessary:
```js
import { _ as _Switch, j as jsx, a as jsxs, r as react, F as Fragment, c as css, R as ReactDOM, b as React } from "./vendor.85adad55.js";
const Switch = _Switch.default;
function Counter() {...}
function FragmentTest() {...}
function App() {
  return /* @__PURE__ */ jsx("div", {
    className: "App",
    children: /* @__PURE__ */ jsxs("header", {
      className: "App-header",
      children: [/* @__PURE__ */ jsx("h1", {
        children: "Hello Vite + React + @emotion/react"
      }), /* @__PURE__ */ jsx(FragmentTest, {}), /* @__PURE__ */ jsxs("p", {
        children: ["Edit ", /* @__PURE__ */ jsx("code", {
          children: "App.jsx"
        }), " and save to test HMR updates."]
      }), /* @__PURE__ */ jsx("a", {
        className: "App-link",
        href: "https://reactjs.org",
        target: "_blank",
        rel: "noopener noreferrer",
        children: "Learn React"
      })]
    })
  });
}
ReactDOM.render(React.createElement(App), document.getElementById("app"));
```

One caveat: This issue does **not** appear when using Emotion.
```js
jsxImportSource: '@emotion/react',
babel: {
  plugins: ['@emotion/babel-plugin']
}
```

aleclarson committed:
refactor: simplify plugin imports

aleclarson committed:
refactor: consolidate `configResolved` hooks

Also:
- check `jsxInject` inside configResolved
- ensure all config values have reasonable defaults for Rollup compatibility

aleclarson committed:
fix: use development version of automatic jsx-runtime transform

aleclarson committed:
fix: ensure jsx source is added to elements in node_modules

aleclarson commented:
Can you make sure this plugin works with ESM Vite project?

pengx17 committed:
fix: tests

pengx17 commented:
> Here's why the `vite:react-jsx` plugin is needed. 👇

Cool! Now I understand why it is being added.

pengx17 commented on packages/playground/react-jsx-runtime/App.jsx:
```diff
@@ -0,0 +1,56 @@
+import { useState } from 'react'
+import { css } from '@emotion/react'
+
+import _Switch from 'react-switch'
```
This line is a bit weird.
The dist in `react-switch` is like the following:

```js
Object.defineProperties(exports, '__esModule', { value: true })
exports.default = ReactSwitch
```

When running with `yarn vite`, `react-switch` will be pre-bundled with esbuild and it will emit `export default xxx` like it is an esm;
but when run with `yarn vite build`, `react-switch` will be bundled with rollup and commonjs plugin, the result is `export default { default: ReactSwitch }`.

I think the way rollup/commonjs plugin works has a bug and it did not correctly handles `__esModule` flag.
This issue might be related: https://github.com/rollup/plugins/issues/939 and the [fix](https://github.com/rollup/plugins/commit/7757aebc74e8b40598e7f67e71dd8e427b0d0095) is included `19.0.2`. I tried this version but still did not correctly transform the default  with esModule flag.

aleclarson replied:
Yeah, there's a Vite issue for this too:
https://github.com/vitejs/vite/issues/2139#issuecomment-854960323

pengx17 replied:
ops, this is an infamous issue already. I think eventually this issue should be fixed on the rollup commonjs plugin. But not sure why it is still not fixed by 19.0.2

githoniel replied:
@pengx17 commonjs@19.0.2 only fix code reassign to `module.exports`  with `__esModule` key like

```
module.exports = { __esModule:  true , default } 
```

it do not change anything when reassign to exports in your case.

your problem is because `react-switch` is not directly export `ReactSwitch`, it's main entry export  in commonjs like below

```js
// main entry
if (process.env.NODE_ENV === "production") {
  module.exports = require("./dist/react-switch.min.js");
} else {
  module.exports = require("./dist/react-switch.dev.js");
}
```

and then export as esm default in `"./dist/react-switch.dev.js"`

Rollup will output code like below in order to match interop pattern to Node.js

```js
var lib = {};

Object.defineProperty(lib, '__esModule', { value: true });
class ReactSwitch {}

lib.default = ReactSwitch;

var demoLib = lib;
```

your may try to set `defaultIsModuleExports: false` in commonjs plugin option since your code will always run in browser.

pengx17 replied:
Wow, thanks for the insights githoniel! This is very helpful!

aleclarson committed:
chore: fix dev build output dir

aleclarson committed:
chore: rename variables

aleclarson committed:
chore: rewrite option description
